lib/rules/no-restricted-imports.js

```diff
@@ -17,33 +27,55 @@ module.exports = {
         },
 
         schema: {
-            type: "array",
-            items: {
-                type: "string"
-            },
-            uniqueItems: true
+            anyOf: [
+                arrayOfStrings,
+                {
+                    type: "array",
+                    items: [{
+                        type: "object",
+                        properties: {
+                            paths: arrayOfStrings,
+                            patterns: arrayOfStrings
+                        },
+                        additionalProperties: false
+                    }],
+                    additionalItems: false
+                }
+            ]
         }
     },
 
     create(context) {
-        const restrictedImports = context.options;
+        const options = Array.isArray(context.options) ? context.options : [];
+        const isStringArray = typeof options[0] !== "object";
+        const restrictedPaths = new Set(isStringArray ? context.options : options[0].paths || []);
+        const restrictedPatterns = isStringArray ? [] : options[0].patterns || [];
 
         // if no imports are restricted we don"t need to check
-        if (restrictedImports.length === 0) {
+        if (restrictedPaths.size === 0 && restrictedPatterns.length === 0) {
             return {};
         }
 
+        const ig = ignore().add(restrictedPatterns);
+
         return {
             ImportDeclaration(node) {
                 if (node && node.source && node.source.value) {
 
-                    const value = node.source.value.trim();
+                    const importName = node.source.value.trim();
 
-                    if (restrictedImports.indexOf(value) !== -1) {
+                    if (restrictedPaths.has(importName)) {
                         context.report(node, "'{{importName}}' import is restricted from being used.", {
-                            importName: value
+                            importName
```

Comment:
While you're in here, can we switch this rule to use new-style reporting? That takes a single object argument and looks like this:

```js
context.report({
    node,
    message: MESSAGE,
    data: { importName }
});
```